File: app/src/main/java/com/example/soroushprofile/prefrences/SettingsPreference.java
package com.example.soroushprofile.prefrences;

import android.graphics.Color;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.lifecycle.ViewModelProvider;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import com.example.soroushprofile.ProfileActivity;
import com.example.soroushprofile.R;
import com.example.soroushprofile.models.ConversationThread;
import com.example.soroushprofile.models.ConversationType;
import com.example.soroushprofile.prefrences.widgets.SoroushPreference;
import com.example.soroushprofile.prefrences.widgets.SoroushPreferenceCategory;
import com.example.soroushprofile.prefrences.widgets.UserMediaPreference;
import com.example.soroushprofile.userprofile.ProfileViewModel;


public class SettingsPreference extends PreferenceFragmentCompat {

    private final static String USER_MEDIA_PREFERENCE = "user_media_preference";
    private final static String USER_MEDIA_PREFERENCE_CATEGORY = "user_media_preference_category";
    private final static String RESTRICT_PREFERENCE = "restrict_preference";
    private final static String DESCRIPTION_PREFERENCE = "description_preference";

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        bindViewModel();
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preferences, rootKey);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initializeRestrictTextColor(findPreference(RESTRICT_PREFERENCE));
    }

    private void bindViewModel() {
        ProfileViewModel model = new ViewModelProvider(requireActivity()).get(ProfileViewModel.class);
        String type = requireActivity().getIntent().getStringExtra(ProfileActivity.CONVERSATION_TYPE);
        model.getConversationThread(type).observe(getViewLifecycleOwner(), this::bindThread);
    }

    private void bindThread(ConversationThread thread) {
        initializeUserMedia(thread);
        initializeDescription(thread);

        switch (thread.getType()) {
            case individual:
                initializeRestrictAction(R.string.profile_user_restrict_individual, thread.getType());
                break;
            case group:
                initializeRestrictAction(R.string.profile_user_restrict_group, thread.getType());
                break;
            case channel:
                initializeRestrictAction(R.string.profile_user_restrict_channel, thread.getType());
                break;
        }
    }


    private void initializeRestrictTextColor(Preference preference) {
        SoroushPreference soroushPreference = (SoroushPreference) preference;
        soroushPreference.setTitleColor(Color.RED);
    }

    private void initializeRestrictAction(@StringRes int title, ConversationType type) {
        Preference blockPreference = findPreference(RESTRICT_PREFERENCE);
        assert blockPreference != null;
        blockPreference.setTitle(title);
        blockPreference.setOnPreferenceClickListener(new BlockClickListener(type));
    }

    private void initializeUserMedia(ConversationThread thread) {
        SoroushPreferenceCategory category = findPreference(USER_MEDIA_PREFERENCE_CATEGORY);
        UserMediaPreference preference = findPreference(USER_MEDIA_PREFERENCE);

        assert preference != null;
        preference.refresh(thread);

        assert category != null;
        category.setOnSummaryClickListener(v -> {
            Toast.makeText(v.getContext(), "More...", Toast.LENGTH_SHORT).show();
        });
        category.setSummary(thread.getMedia().isEmpty()
                ? null : getString(R.string.profile_user_media_category_more_title));
        category.setTitle(thread.getMedia().isEmpty()
                ? null : getString(R.string.profile_user_media_category_title));
    }

    private void initializeDescription(ConversationThread thread) {
        Preference preference = findPreference(DESCRIPTION_PREFERENCE);
        assert preference != null;

        if (thread.getType().isIndividual()) {
            preference.setSummary(R.string.profile_user_description_bio_title);
            preference.setTitle(thread.getDescription().orElse(getString(R.string.profile_user_description_bio_empty)));
        } else {
            preference.setSummary(R.string.profile_user_description_title);
            preference.setTitle(thread.getDescription().orElse(getString(R.string.profile_user_description_empty)));
        }

        preference.setSummary(thread.getType().isIndividual()
                ? R.string.profile_user_description_bio_title : R.string.profile_user_description_title);
    }

    private class BlockClickListener implements Preference.OnPreferenceClickListener {

        private final ConversationType type;

        private BlockClickListener(ConversationType type) {
            this.type = type;
        }

        @Override
        public boolean onPreferenceClick(Preference preference) {
            switch (type) {
                case individual:
                    Toast.makeText(getContext(), "Block Action Trigger...", Toast.LENGTH_SHORT).show();
                    break;
                case group:
                    Toast.makeText(getContext(), "Leave Group Trigger...", Toast.LENGTH_SHORT).show();
                    break;
                case channel:
                    Toast.makeText(getContext(), "Leave Channel Trigger...", Toast.LENGTH_SHORT).show();
                    break;
            }
            return true;
        }

    }
}

File: app/src/main/java/com/example/soroushprofile/prefrences/SettingsPreference.kt
package com.example.soroushprofile.prefrences

import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.annotation.StringRes
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat

import com.example.soroushprofile.ProfileActivity
import com.example.soroushprofile.R
import com.example.soroushprofile.models.ConversationThread
import com.example.soroushprofile.models.ConversationType
import com.example.soroushprofile.prefrences.widgets.SoroushPreference
import com.example.soroushprofile.prefrences.widgets.SoroushPreferenceCategory
import com.example.soroushprofile.prefrences.widgets.UserMediaPreference
import com.example.soroushprofile.userprofile.ProfileViewModel


class SettingsPreference : PreferenceFragmentCompat() {

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        bindViewModel()
    }

    override fun onCreatePreferences(savedInstanceState: Bundle, rootKey: String) {
        setPreferencesFromResource(R.xml.preferences, rootKey)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initializeRestrictTextColor(findPreference(RESTRICT_PREFERENCE))
    }

    private fun bindViewModel() {
        val model = ViewModelProvider(requireActivity()).get(ProfileViewModel::class.java)
        val type = requireActivity().intent.getStringExtra(ProfileActivity.CONVERSATION_TYPE)
        model.getConversationThread(type).observe(viewLifecycleOwner, Observer<ConversationThread> { this.bindThread(it) })
    }

    private fun bindThread(thread: ConversationThread) {
        initializeUserMedia(thread)
        initializeDescription(thread)

        when (thread.type) {
            ConversationType.Individual -> initializeRestrictAction(R.string.profile_user_restrict_individual, thread.type)
            ConversationType.Group -> initializeRestrictAction(R.string.profile_user_restrict_group, thread.type)
            ConversationType.Channel -> initializeRestrictAction(R.string.profile_user_restrict_channel, thread.type)
        }
    }


    private fun initializeRestrictTextColor(preference: Preference?) {
        val soroushPreference = preference as SoroushPreference?
        soroushPreference!!.setTitleColor(Color.RED)
    }

    private fun initializeRestrictAction(@StringRes title: Int, type: ConversationType) {
        val blockPreference = findPreference<Preference>(RESTRICT_PREFERENCE)!!
        blockPreference.setTitle(title)
        blockPreference.onPreferenceClickListener = BlockClickListener(type)
    }

    private fun initializeUserMedia(thread: ConversationThread) {
        val category = findPreference<SoroushPreferenceCategory>(USER_MEDIA_PREFERENCE_CATEGORY)
        val preference = findPreference<UserMediaPreference>(USER_MEDIA_PREFERENCE)!!

        preference.refresh(thread)

        category!!.setOnSummaryClickListener(View.OnClickListener { v: View? -> Toast.makeText(v!!.context, "More...", Toast.LENGTH_SHORT).show() })
        category.summary = if (thread.media!!.isEmpty())
            null
        else
            getString(R.string.profile_user_media_category_more_title)
        category.title = if (thread.media!!.isEmpty())
            null
        else
            getString(R.string.profile_user_media_category_title)
    }

    private fun initializeDescription(thread: ConversationThread) {
        val preference = findPreference<Preference>(DESCRIPTION_PREFERENCE)!!

        if (thread.type.isIndividual) {
            preference.setSummary(R.string.profile_user_description_bio_title)
            preference.setTitle(thread.description ?: getString(R.string.profile_user_description_bio_empty))
        } else {
            preference.setSummary(R.string.profile_user_description_title)
            preference.setTitle(thread.description ?: getString(R.string.profile_user_description_empty))
        }

        preference.setSummary(if (thread.type.isIndividual)
            R.string.profile_user_description_bio_title
        else
            R.string.profile_user_description_title)
    }

    private inner class BlockClickListener public constructor(private val type: ConversationType) : Preference.OnPreferenceClickListener {

        override fun onPreferenceClick(preference: Preference): Boolean {
            when (type) {
                ConversationType.Individual -> Toast.makeText(context, "Block Action Trigger...", Toast.LENGTH_SHORT).show()
                ConversationType.Group -> Toast.makeText(context, "Leave Group Trigger...", Toast.LENGTH_SHORT).show()
                ConversationType.Channel -> Toast.makeText(context, "Leave Channel Trigger...", Toast.LENGTH_SHORT).show()
            }
            return true
        }

    }

    companion object {

        private val USER_MEDIA_PREFERENCE = "user_media_preference"
        private val USER_MEDIA_PREFERENCE_CATEGORY = "user_media_preference_category"
        private val RESTRICT_PREFERENCE = "restrict_preference"
        private val DESCRIPTION_PREFERENCE = "description_preference"
    }
}
